demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c

```diff
@@ -0,0 +1,1608 @@
+ */
+#define MQTT_EXAMPLE_TOPIC                  CLIENT_IDENTIFIER "/example/topic"
+
+/**
+ * @brief Length of client MQTT topic.
+ */
+#define MQTT_EXAMPLE_TOPIC_LENGTH           ( ( uint16_t ) ( sizeof( MQTT_EXAMPLE_TOPIC ) - 1 ) )
+
+/**
+ * @brief The MQTT message published in this example.
+ */
+#define MQTT_EXAMPLE_MESSAGE                "Hello World!"
+
+/**
+ * @brief The MQTT message published in this example.
+ */
+#define MQTT_EXAMPLE_MESSAGE_LENGTH         ( ( uint16_t ) ( sizeof( MQTT_EXAMPLE_MESSAGE ) - 1 ) )
+
+/**
+ * @brief Maximum number of outgoing publishes maintained in the application
+ * until an ack is received from the broker.
+ */
+#define MAX_OUTGOING_PUBLISHES              ( 5U )
+
+/**
+ * @brief Invalid packet identifier for the MQTT packets. Zero is always an
+ * invalid packet identifier as per MQTT 3.1.1 spec.
+ */
+#define MQTT_PACKET_ID_INVALID              ( ( uint16_t ) 0U )
+
+/**
+ * @brief Timeout for MQTT_ProcessLoop function in milliseconds.
+ *
+ */
+#define MQTT_PROCESS_LOOP_TIMEOUT_MS        ( 500U )
+
+/**
+ * @brief Time interval in seconds at which an MQTT PINGREQ need to be sent to
+ * broker.
+ */
+#define MQTT_KEEP_ALIVE_INTERVAL_SECONDS    ( 60U )
+
+/**
+ * @brief Delay between MQTT publishes in seconds.
+ */
+#define DELAY_BETWEEN_PUBLISHES_SECONDS     ( 1U )
+
+/**
+ * @brief Transport timeout in milliseconds for transport send and receive.
+ */
+#define TRANSPORT_SEND_RECV_TIMEOUT_MS      ( 20 )
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Structure to keep the MQTT publish packets until an ack is received
+ * for QoS1 publishes.
+ */
+typedef struct PublishPackets
+{
+    /**
+     * @brief Packet identifier of the publish packet.
+     */
+    uint16_t packetId;
+
+    /**
+     * @brief Publish info of the publish packet.
+     */
+    MQTTPublishInfo_t pubInfo;
+} PublishPackets_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief globalEntryTime entry time into the application to use as a reference
+ * timestamp in #getTimeMs function. #getTimeMs will always return the difference
+ * of current time with the globalEntryTime. This will reduce the chances of
+ * overflow for 32 bit unsigned integer used for holding the timestamp.
+ *
+ */
+static uint32_t globalEntryTimeMs = 0U;
+
+/**
+ * @brief Packet Identifier generated when Subscribe request was sent to the broker;
+ * it is used to match received Subscribe ACK to the transmitted subscribe.
+ */
+static uint16_t globalSubscribePacketIdentifier = 0U;
+
+/**
+ * @brief Packet Identifier generated when Unsubscribe request was sent to the broker;
+ * it is used to match received Unsubscribe response to the transmitted unsubscribe
+ * request.
+ */
+static uint16_t globalUnsubscribePacketIdentifier = 0U;
+
+/**
+ * @brief Array to keep the outgoing publish messages.
+ * These stored outgoing publish messages are used to keep the messages until
```

Comment:
These stored outgoing publish messages are kept until a successful ack is received.